src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/MLKemImplementation.OpenSsl.cs

```diff
@@ -0,0 +1,123 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using Microsoft.Win32.SafeHandles;
+
+namespace System.Security.Cryptography
+{
+    internal sealed class MLKemImplementation : MLKem
+    {
+        private SafeEvpPKeyHandle _key;
+
+        // OpenSSL is expected to give "all or none" support.
+        internal static new bool IsSupported => Interop.Crypto.EvpKemAlgs.MlKem512 is not null;
+
+        private MLKemImplementation(MLKemAlgorithm algorithm, SafeEvpPKeyHandle key) : base(algorithm)
+        {
+            _key = key;
+        }
+
+        internal static MLKem GenerateKeyImpl(MLKemAlgorithm algorithm)
+        {
+            Debug.Assert(IsSupported);
+            SafeEvpKemHandle handle = MapAlgorithmToHandle(algorithm); // Shared handle, do not dispose.
```

Comment:
Is it necessary to have this handle? It looks like it's mainly just used in key generation/import native code to get the name back, so couldn't we do the mapping on the C# side and pass in the name directly instead?

Reply:
I can get rid of the handle in the case of ML-KEM. The important part is correctly identifying if the algorithm is supported or not via fetch, but we don't need to keep the handle for that. We can just introduce a map.